idaes/apps/uncertainty_propagation/uncertainties.py

```diff
@@ -0,0 +1,248 @@
+    # Remove all "'" and " " in theta_names
+    theta_names, var_dic,variable_clean = clean_variable_name(theta_names)
+    parmest_class = parmest.Estimator(model_function, data, theta_names, obj_function,
+                 tee, diagnostic_mode, solver_options)
+    obj, theta,cov = parmest_class.theta_est(calc_cov=True)
+    # Convert theta keys to the original name 
+    # Revert theta_names to be original
+    if variable_clean:
+        theta_out = {}
+        for i in var_dic.keys():
+            theta_out[var_dic[i]] = theta[i]
+        theta_names = [var_dic[v] for v in theta_names] 
+    else:
+        theta_out = theta
+    propagate_results  =  propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee)
+   
+    if len(propagate_results.gradient_c)>0:
+        I,J,V =  propagate_results.gradient_c[:,1].flatten().astype(int)-1, propagate_results.gradient_c[:,0].flatten().astype(int)-1, propagate_results.gradient_c[:,2].flatten()
+        coo_ = sparse.coo_matrix((V,(I,J)))
+    else:
+        coo_ = propagate_results.gradient_c
+
+    Output = namedtuple('Output',['obj', 'theta', 'theta_names', 'cov','gradient_f', 'gradient_c', 'dsdp', 'propagation_c', 'propagation_f','col'])
+    results= Output(obj, theta_out, theta_names, cov, 
+                     propagate_results.gradient_f,
+                     propagate_results.gradient_c, 
+                     propagate_results.dsdp,
+                     coo_,
+                     propagate_results.propagation_f,
+                     propagate_results.col)
+    return results
+
+def propagate_uncertainty(model_uncertain, theta, cov, theta_names, tee=False, solver_options=None):
+    """
+    This function calculates gradient vector, expectation, and variance of the objective function and constraints  of the model for given estimated optimal parameters and covariance matrix of parameters. It calculates error propagation of the objective function and constraints by using gradient vector and covariance matrix.
+    
+    Args:
+        model_uncertain(function or Pyomo ConcreteModel) : Function is a a python Function that generates an instance of the Pyomo model
+        theta(dict)                    : Estimated parameters 
+        cov(numpy.ndarray)             : Covariance matrix of parameters 
+        theta_names(list of strings)   : List of estimated Var names
+        var_dic(dictionary)            : If any original variable contains apostrophes, need an auxiliary dictionary with keys theta_namess without apostrophes, values with apostrophes. e.g) var_dic: {'fs.properties.tau[benzene,toluene]': "fs.properties.tau['benzene','toluene']", 'fs.properties.tau[toluene,benzene]': "fs.properties.tau['toluene','benzene']"} 
+        tee(bool, optional)            : Indicates that ef solver output should be teed
+        solver_options(dict, optional) : Provides options to the solver (also the name of an attribute)
+
```

Comment:
Add a few lines here defining:
Ncon = number of constraints
Ntheta = number of uncertainty parameters
Nx = number of decision (primal) variables.

Then in the doc string for each element of results, give the dimensions, e.g., Nx by 1 vector or Ncon by Ntheta matrix. This will really help with clarity.

Reply:
updated it.